### PR TITLE
Geoip2 lookup individual values

### PR DESCRIPTION
**Pull request by neilcook**
This PR implements Lua functions to lookup individual values, to allow any GeoIP DB to be interrogated.

### CODE REVIEW

File: common/wforce-geoip2.cc [hunk 1]
```diff
@@ -156,7 +156,7 @@ bool WFGeoIP2DB::lookupDataValue(const ComboAddress& address, const std::vector<
     }
     for (auto& i : path) {
       if (i != nullptr)
-        delete i;
+        delete[] i;
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: iirc delete[] is NULL safe.

File: common/wforce-geoip2.cc [hunk 2]
```diff
@@ -135,6 +135,93 @@ bool WFGeoIP2DB::mmdbLookup(const std::string& ip, MMDB_lookup_result_s& res)
   return false;
 }
 
+char* convertToCharStar(const std::pair<unsigned int, std::string>& pair)
+{
+  char *cs = new char[pair.second.size()+1];
+  std::strcpy(cs, pair.second.c_str());
+  return cs;
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: `return std::strdup(pair.second.c_str())`?

File: common/wforce-geoip2.cc [hunk 3]
```diff
@@ -135,6 +135,93 @@ bool WFGeoIP2DB::mmdbLookup(const std::string& ip, MMDB_lookup_result_s& res)
...
+  
+  if (mmdbLookup(address.toString(), res)) {
+    std::transform(attrs.begin(), attrs.end(), std::back_inserter(path), convertToCharStar);
+    path.emplace_back(nullptr);
+    if ((MMDB_aget_value(&res.entry, &ret_data, &path[0]) == MMDB_SUCCESS) && (ret_data.has_data)) {
+      retval = true;
+    }
+    for (auto& i : path) {
+      if (i != nullptr)
+        delete i;
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: delete is null safe, also you probably should use delete[] here?

File: common/wforce-geoip2.cc (hunk 2 above)
Comment (neilcook) [reply common/wforce-geoip2.cc]: Yeah that would also work. Also I guess I'd need to use free() rather than delete[] (and keep the check for nullptr)

File: common/wforce-geoip2.cc (hunk 1 above)
Comment (neilcook) [reply common/wforce-geoip2.cc]: Weirdly I only added the check because I got a SEGV otherwise

File: common/wforce-geoip2.cc [hunk 4]
```diff
@@ -156,7 +155,7 @@ bool WFGeoIP2DB::lookupDataValue(const ComboAddress& address, const std::vector<
     }
     for (auto& i : path) {
       if (i != nullptr)
-        delete[] i;
+        free(i);
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: free is null-safe too.

Comment (neilcook) [reply common/wforce-geoip2.cc]: well I never!

File: common/wforce-geoip2.cc (hunk 3 above)
Comment (cmouse) [reply common/wforce-geoip2.cc]: no free here?

File: docker/Makefile.am [hunk 5]
```diff
@@ -12,7 +12,7 @@ $(GEOIP_FILE_GZ):
 $(GEOIP_FILE): $(GEOIP_FILE_GZ)
 	gunzip -c $(GEOIP_FILE_GZ) >$(GEOIP_FILE)
 
-$(COMPOSE_TARGET): $(COMPOSE_SOURCE) $(GEOIP_FILE) $(shell find ../common -type f) $(shell find ../wforce -type f) $(shell find ../trackalert -type f) $(shell find ../ext -type f) 
+$(COMPOSE_TARGET): $(COMPOSE_SOURCE) $(GEOIP_FILE) $(shell find ../common -type f) $(shell find ../wforce -type f) $(shell find ../trackalert -type f) $(shell find ../ext -type f) $(shell find ../regression-tests -type f)
```
Comment (cmouse) [inline docker/Makefile.am]: why is this here?

File: common/wforce-geoip2.cc [hunk 6]
```diff
@@ -156,7 +155,7 @@ bool WFGeoIP2DB::lookupDataValue(const ComboAddress& address, const std::vector<
     }
     for (auto& i : path) {
       if (i != nullptr)
-        delete i;
+        free(i);
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: this could already be done in previous commit instead of first doing it there and then changing it here.

File: common/wforce-geoip2.cc [hunk 7]
```diff
@@ -137,9 +138,7 @@ bool WFGeoIP2DB::mmdbLookup(const std::string& ip, MMDB_lookup_result_s& res)
 
 char* convertToCharStar(const std::pair<unsigned int, std::string>& pair)
 {
-  char *cs = new char[pair.second.size()+1];
-  std::strcpy(cs, pair.second.c_str());
-  return cs;
+  return strdup(pair.second.c_str());
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: this could already be done in the earlier commit.

File: common/wforce-geoip2.cc [hunk 8]
```diff
@@ -154,8 +154,7 @@ bool WFGeoIP2DB::lookupDataValue(const ComboAddress& address, const std::vector<
       retval = true;
     }
     for (auto& i : path) {
-      if (i != nullptr)
-        free(i);
+      free(i);
```
Comment (cmouse) [inline common/wforce-geoip2.cc]: this could already happen in earlier commit.
